contracts/MetawoodAuction.sol

```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "./interfaces/IMetawoodNFT.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC1155/utils/ERC1155Holder.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+contract MetawoodAuction is ERC1155Holder, Ownable, ReentrancyGuard {
+    using SafeMath for uint256;
+    using Counters for Counters.Counter;
+
+    Counters.Counter private _auctionCounter;
+    IMetawoodNFT public metawoodNFT;
+
+    uint256 public bidIncreaseThreshold = 200;
+
+    constructor(address _nftContract) {
+        metawoodNFT = IMetawoodNFT(_nftContract);
+    }
+
+    enum AuctionState {
+        CLOSED,
+        ACTIVE
+    }
+
+    //single struct for handling bid, tbd?
+    struct Auction {
+        uint256 id;
+        IMetawoodNFT nftContract;
+        uint256 nftId; // NFT Id
+        address payable creator; // Creator of the Auction
+        uint256 bidIncreaseThreshold;
+        uint256 auctionEnd;
+        uint256 minimumBid;
+        uint256 highestBid;
+        address payable highestBidder;
+        uint256[] bids;
```

Comment:
Not sure around this approach.
What if an attack is just an attacker bidding continuously for his own NFT with a very less auction price from different address.
He can do 100txs (quite possible with less gas on MATIC) and the contract has to store all the biddings.
Think to store only the latest bid. Might make things easier for you.
Backend can store all the tx history in some way
Please check https://etherscan.io/address/0xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb#code
Its for cryptopunks. They also store the highest bid only

Reply:
This one was kept for the initial code flow for the bid to be revoked as well, initially, there was an option of canceling the bid and at that time we needed to keep the last bidder and its bid amount escrowed since we had to readjust the max bid if highest bid was canceled, but in a call, we discussed that we don't allow bid cancellation and this was left, this had another purpose to directly get the bidding amounts for price graph